### PR TITLE
[Perf] Optimize Cosmos3 SP gather and video encode path

### PR DESCRIPTION
**Pull request by david6666666**
## Summary

This draft stacks on the Cosmos3 postprocess optimization from #4615 and targets three remaining framework overheads seen when comparing vLLM-Omni against A-framework on Cosmos3-Nano:

1. reduce video-only sequence-parallel gather payload after the GEN blocks;
2. avoid Cosmos3 tensor MP4 encoding range-probe synchronizations after the #4615 unit-interval tensor fast path;
3. use a Cosmos3-only pinned host transfer path for the final tensor-to-NumPy copy before video encoding.

The VAE internal parallel decode port and denoise micro-fusion experiments are intentionally not included here because they need separate correctness gates or did not improve latency.

## Optimization Points

| Area | Baseline finding | Change in this PR | Expected effect | Status |
|---|---:|---|---|---|
| GEN SP gather payload | vLLM profile5 copy/layout avg ~1704.8 ms/rank vs A-framework ~219.4 ms | In the Cosmos3 video-only path, run final RMSNorm + `proj_out` before `gen_sp_gather` so the gather moves `patch_latent_dim` tokens instead of hidden-size GEN states | Lower SP gather payload and downstream copy/layout pressure | Unit-tested and 35-step benchmarked |
| Video response encode sync | vLLM rank0 profile shows `aten::copy_`/`aten::to`/`_to_copy` plus sync-heavy encode path; DtoH observed in profile | For Cosmos3 floating tensor outputs, skip `min().item()` / `max().item()` range probes because #4615 already clamps no-guardrails tensor output to `[0, 1]` | Remove two forced GPU sync points in serving encode | Unit/API tested and 35-step benchmarked |
| Final video DtoH transfer | vLLM profile shows rank0 pageable DtoH spike; serving path still used pageable `.cpu().numpy()` | For Cosmos3 tensor outputs, copy the final contiguous video tensor into pinned host memory with `non_blocking=True`, synchronize once, and return the NumPy view; fallback to `.cpu().numpy()` on unsupported paths | Reduce final host transfer cost before MP4 encode | Unit/API tested and isolated B300 microbenchmarked |
| VAE decode layout | vLLM `vae.decode` remains slower with CUDNN layout transforms prominent | No change included | Needs decoder-internal height shard + layer-local halo exchange; outer-shard shortcut drifted raw decoded tensor | Rejected for this PR |
| Denoise micro-fusion | Single-element `torch.cat` in video-only path looked removable | No change included | 5-step profile showed no latency improvement | Rejected for this PR |

## Benchmark

B300 GPUs `3,4,5,6`, `nvidia/Cosmos3-Nano`, CUDNN attention backend, `cfg_parallel_size=2`, `ulysses_degree=2`, `vae_patch_parallel_size=4`, same reconstructed 35-step request for base #4615 and the first two changes in this PR.

| Run | Commit | Steps | Wall | Inference | `stage_0_gen_ms` | Output |
|---|---|---:|---:|---:|---:|---|
| #4615 base | `a4347e0b` | 35 | 14.078s | 13.046s | 12092.180ms | `sha256=e1509d67...`, 4,393,353 bytes |
| SP gather + range-sync skip | `493557e` | 35 | 12.429s | 12.361s | 11629.336ms | `sha256=e1509d67...`, 4,393,353 bytes |
| Delta | | | **-1.649s / +11.71%** | **-0.686s** | **-462.844ms / +3.83%** | identical |

After the pinned-host-transfer commit, a full 4-GPU API rerun was completed for the same `seconds=4` default-frame request shape. It showed no material latency regression versus the previous branch benchmark, but the independent full generation output was not byte/frame-md5 identical across server runs, so correctness for the transfer-only patch is still gated by the isolated tensor-copy equality test and unit/API tests.

| Case | Tensor | Pageable `.cpu().numpy()` | Pinned host transfer | Correctness |
|---|---|---:|---:|---|
| Final video tensor transfer microbench | `189x720x1280x3 uint8` | 155.78ms avg | 9.28ms avg | arrays equal; hash unchanged |

## Validation

| Check | Result |
|---|---|
| `git diff --check HEAD~3..HEAD` | passed |
| `python -m compileall` on touched source/tests | passed |
| `ruff check` on touched source/tests | passed |
| `pytest tests/diffusion/models/cosmos3/test_cosmos3_transformer.py -q` | 19 passed |
| `pytest tests/entrypoints/openai_api/test_video_api_utils.py tests/entrypoints/openai_api/test_video_server.py -q` | 77 passed |
| fixed 4-GPU Cosmos3-Nano 35-step benchmark for the first two changes | passed; output identical |
| isolated B300 final DtoH microbench for pinned transfer | passed; arrays equal |
| fixed 4-GPU Cosmos3-Nano 35-step benchmark after pinned transfer | completed; wall 12.469s, stage 11635.208ms for default 93-frame `seconds=4` request; independent output hash/frame-md5 differed, so not used as correctness proof |

Artifacts on the B300 host:

| Artifact | Path |
|---|---|
| #4615 same-request 35-step | `/home/zjy/code/david/tmp/cosmos3_pr4615_same_req_steps35_20260623_015225` |
| SP gather + range-sync same-request 35-step | `/home/zjy/code/david/tmp/cosmos3_pr4629_same_req_steps35_20260623_015335` |
| VAE internal decode audit | `/home/zjy/code/david/tmp/cosmos3_vae_internal_decode_port_audit_20260623_1730/vae_internal_decode_port_audit.md` |
| VAE rejected report | `/home/zjy/code/david/tmp/cosmos3_vae_internal_decode_port_rejected_20260623_worker.md` |
| Denoise micro-fusion rejected report | `/home/zjy/code/david/tmp/cosmos3_denoise_layout_fusion_opt_rejected_20260623.md` |
| Final pinned-transfer 93-frame async run | `/home/zjy/code/david/tmp/cosmos3_pr4631_pinned_same_req_steps35_20260623_095540` |
| Final pinned-transfer 189-frame async run | `/home/zjy/code/david/tmp/cosmos3_pr4631_pinned_same_req_steps35_20260623_095219` |

## Risk

- The SP gather change is limited to the Cosmos3 video-only path. Action, sound, and transfer-control paths keep the existing gather-before-split behavior.
- The range-probe skip and pinned host transfer are gated to Cosmos3 tensor outputs. Other video models keep the default encoding path.
- Pinned host transfer falls back to `.cpu().numpy()` if pinned allocation or async copy is unsupported.
- The final pinned-transfer commit has now been run in a full 4-GPU API request. The timing is stable, but full-generation MP4 hash/frame-md5 differed across independent server runs, so final review should rely on unit/API tests and the isolated tensor-copy equality test for this transfer-only change, or add raw tensor instrumentation if strict full-output determinism is required.

### CODE REVIEW

**Comment by david6666666** [comment]
Optimization/validation summary:

| Optimization point | Included in PR | Baseline finding | Evidence from this PR | Current validation |
|---|---|---|---|---|
| Cosmos3 video-only SP gather payload reduction | Yes | A-framework comparison showed vLLM copy/layout pressure was much higher in the profiled path | Same 35-step request improved wall by 1.649s and `stage_0_gen_ms` by 462.844ms with identical MP4 output | Cosmos3 transformer tests and 35-step serving benchmark passed |
| Cosmos3 tensor encode range-probe sync removal | Yes | vLLM rank0 profile showed sync-heavy copy/layout and response encoding work before MP4 bytes were returned | Same 35-step request reduced MP4 response encoding from 927.36ms to 704.13ms while keeping identical output hash | video API/server tests passed; output hash identical in benchmark |
| VAE decode layout | No | VAE decode remains a separate bottleneck with CUDNN layout transforms | Not included because tested VAE layout/chunk candidates either had no clear gain or changed output bytes | Kept out of this PR intentionally |
| Denoise communication packing | No | Denoise gap also includes communication and launch fragmentation | Covered by a separate draft PR | Kept separate to reduce review risk |

Overall: this PR is the current low-risk framework/postprocess slice. It preserves output bytes for the fixed Cosmos3-Nano request while reducing wall time, stage time, and MP4 response encoding time, and it avoids mixing in VAE or attention-communication changes that need separate review.

File: tests/diffusion/models/cosmos3/test_cosmos3_transformer.py [hunk 1]
```diff
@@ -163,16 +163,18 @@ def test_forward_accepts_transfer_control_latents(monkeypatch: pytest.MonkeyPatc
...
 
-def test_forward_gathers_gen_tokens_before_unpatchify(monkeypatch: pytest.MonkeyPatch) -> None:
+def test_forward_gathers_patch_tokens_before_unpatchify(monkeypatch: pytest.MonkeyPatch) -> None:
     from vllm_omni.diffusion.models.cosmos3 import transformer_cosmos3
 
     class RecordingGather(nn.Module):
         def __init__(self) -> None:
             super().__init__()
             self.calls = 0
+            self.last_shape: tuple[int, ...] | None = None
```
Comment (hsliuustc0106) [inline tests/diffusion/models/cosmos3/test_cosmos3_transformer.py:173]: Good test - verifying the gather shape is `patch_latent_dim` confirms the optimization reduces payload size.

**Review by hsliuustc0106 — COMMENTED** [review]
1 comment above

**Comment by david6666666** [comment]
Updated this draft with a third serving-side optimization: Cosmos3-only pinned host transfer for the final video tensor before MP4 encoding.

| Optimization point | Scope | Evidence | Result |
|---|---|---|---|
| Video-only SP gather after projection | Transformer GEN video-only path | 35-step same request vs #4615 | Wall 14.078s -> 12.429s; stage 12092.180ms -> 11629.336ms; output hash identical |
| Skip encode range probes | Cosmos3 floating tensor encode path | Same 35-step benchmark and unit/API tests | Removes two scalar sync probes; included in the identical-output 35-step run |
| Pinned final DtoH transfer | Cosmos3 tensor-to-NumPy path before MP4 encode | Isolated B300 final tensor microbench, `189x720x1280x3 uint8` | Pageable 155.78ms avg -> pinned 9.28ms avg; arrays equal/hash unchanged |
| VAE internal decode | Separate VAE path | Source audit + rejected outer-shard raw gate | Not included; requires decoder-internal height shard with layer-local halo exchange |
| Denoise micro-fusion | Single-element `torch.cat` bypass | 5-step profile | Not included; output identical but no latency improvement |

Validation after stacking the pinned-transfer commit onto this branch:

| Check | Result |
|---|---|
| `git diff --check HEAD~3..HEAD` | passed |
| `python -m compileall` touched source/tests | passed |
| `ruff check` touched source/tests | passed |
| `pytest tests/diffusion/models/cosmos3/test_cosmos3_transformer.py -q` | 19 passed |
| `pytest tests/entrypoints/openai_api/test_video_api_utils.py tests/entrypoints/openai_api/test_video_server.py -q` | 77 passed |

Note: the pinned-transfer commit has isolated transfer evidence but still needs a full 4-GPU same-request API benchmark before this draft should be marked ready.

**Comment by david6666666** [comment]
Follow-up validation after stacking the pinned host-transfer commit (`40ff6244`) onto this branch.

| Run | Commit | Request shape | Wall | Inference | `stage_0_gen_ms` | MP4 encode | Output check |
|---|---|---|---:|---:|---:|---:|---|
| Earlier branch benchmark | `493557e` | `seconds=4`, default frames: 93 | 12.429s | 12.361s | 11629.336ms | ~704ms from prior logs | MP4 hash matched the #4615 comparison run |
| Final stacked head | `40ff6244` | `seconds=4`, default frames: 93 | 12.469s | 12.388s | 11635.208ms | 728.91ms | Independent rerun produced 93 frames; MP4 hash/frame-md5 differed, so this is not a same-output proof |
| Final stacked head, true 189-frame check | `40ff6244` | `seconds=4`, `num_frames=189` | 32.320s | 31.263s | 29961.851ms | 1269.33ms | Completed successfully; separate request shape from the earlier 93-frame benchmark |

Notes:

- The 93-frame final-head run shows no material latency regression from the pinned-transfer commit: wall +39.8ms and stage +5.9ms versus the previous branch benchmark, within run noise.
- The full generation output is not deterministic enough across independent server runs to use MP4 hash/frame-md5 as the correctness gate for this final transfer-only patch.
- The pinned-transfer correctness gate remains the isolated final tensor copy microbench plus unit/API tests: `189x720x1280x3 uint8` pageable copy 155.78ms avg -> pinned copy 9.28ms avg, arrays equal/hash unchanged.
- I also found that the earlier branch benchmark used the default `seconds=4` frame count of 93 frames. Explicit `num_frames=189` is a different, heavier request shape and should be reported separately.

Artifacts:

| Artifact | Path |
|---|---|
| Final 93-frame async run | `/home/zjy/code/david/tmp/cosmos3_pr4631_pinned_same_req_steps35_20260623_095540` |
| Final 189-frame async run | `/home/zjy/code/david/tmp/cosmos3_pr4631_pinned_same_req_steps35_20260623_095219` |

**Comment by david6666666** [comment]
Updated 5-step profile comparison after the latest host-transfer change.

Request shape: Cosmos3 Nano I2V serving profile, 5 diffusion steps, 189 frames, 720x1280, same prompt/image/seed and 4-GPU SP profile setup. Operator numbers below are torch-profiler averages per rank.

| Variant | forward | GPU active | CPU bubble | attention | GEMM | copy/layout | elem/act | NCCL | norm | VAE conv | upsample |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| #4615 baseline | 8361.5 ms | 7147.7 ms | 1213.8 ms | 1896.1 ms | 1891.5 ms | 1679.2 ms | 1073.0 ms | 440.4 ms | 125.0 ms | 60.4 ms | 29.4 ms |
| #4630 comm packing | 8201.6 ms | 7086.2 ms | 1115.4 ms | 1886.1 ms | 1883.5 ms | 1673.9 ms | 1068.7 ms | 411.4 ms | 120.3 ms | 60.5 ms | 29.3 ms |
| this PR | 8203.2 ms | 7078.3 ms | 1125.0 ms | 1890.1 ms | 1884.8 ms | 1610.2 ms | 1065.1 ms | 462.5 ms | 119.9 ms | 60.5 ms | 29.3 ms |

Delta vs #4615:

| Variant | forward | GPU active | CPU bubble | attention | GEMM | copy/layout | elem/act | NCCL | norm |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| #4630 comm packing | -159.9 ms | -61.6 ms | -98.3 ms | -9.9 ms | -8.1 ms | -5.2 ms | -4.3 ms | -29.0 ms | -4.7 ms |
| this PR | -158.2 ms | -69.5 ms | -88.8 ms | -5.9 ms | -6.7 ms | -69.0 ms | -7.9 ms | +22.1 ms | -5.0 ms |

Notes:

- This PR primarily reduces the postprocess/gather/host-transfer side: copy/layout improves by about 69 ms/rank and CPU bubble improves by about 89 ms/rank vs #4615 in this 5-step profile.
- #4630 primarily reduces communication: NCCL improves by about 29 ms/rank vs #4615. This PR does not include #4630's Q/K/V all-to-all packing, so NCCL is expectedly higher here.
- The 5-step profile output SHA for this PR matches #4630 for the same request: `5bb3b6bb4b58d19e55e220f8b69d414d428912f87f15b23b7dcd39e68f6bafbb`.
- The remaining largest gaps to the A-framework reference are still copy/layout and CPU bubble, not attention alone: this PR is still around 1.61 s/rank in copy/layout and 1.12 s/rank in CPU bubble.

Recommended next steps:

| Priority | Target | Reason | Suggested path |
| --- | --- | --- | --- |
| P0 | Stack this PR with #4630 | #4630 and this PR improve different buckets, and the combined branch has already passed compile/ruff/API/SP smoke tests | Rebase one PR on the other and run a full 35-step plus 5-step profile on the stacked branch |
| P0 | Layout/copy cleanup | copy/layout is still the largest non-attention gap | Keep tensors in backend-preferred layout across the transformer/VAE boundary and remove repeated contiguous/permute conversions |
| P0 | Host sync cleanup | CPU bubble remains much higher than the A-framework reference | Continue removing blocking `.item()`, pageable D2H, and status/postprocess synchronization points |
| P1 | VAE decoder internal spatial shard | Outer sharding showed speed potential but had decoded drift | Only pursue a decoder-internal shard with layer-local halo exchange and strict precision validation |

Profile artifacts are kept on the remote worker under `/home/zjy/code/david/tmp/cosmos3_pr4631_profile5_split_20260623_100351/`.

**Comment by david6666666** [comment]
Follow-up combined stack validation for latest this PR + #4630.

Temporary stack:

- latest this PR: `40ff6244`
- #4630 communication packing: `009ce14a`
- combined head tested: `c660bbed`

5-step profiler request, 189 frames:

| Variant | wall | inference | stage | output SHA |
| --- | ---: | ---: | ---: | --- |
| #4615 baseline | 13.770s | n/a | 10964.285 ms | `d418c37e...` |
| #4630 only | 11.646s | n/a | 10113.625 ms | `5bb3b6bb...` |
| this PR only | 11.388s | 11.365s | 10072.056 ms | `5bb3b6bb...` |
| this PR + #4630 | 11.350s | 11.327s | 10047.289 ms | `5bb3b6bb...` |

Warm 35-step serving request, default 93 frames:

| Variant | wall | inference | stage | MP4 encode | output SHA |
| --- | ---: | ---: | ---: | ---: | --- |
| #4615 baseline | 14.078s | 13.046s | 12092.180 ms | n/a | `e1509d67...` |
| #4630 only | 13.761s | 12.725s | 11676.404 ms | n/a | `e1509d67...` |
| this PR final | 12.469s | 12.388s | 11635.208 ms | 728.91 ms | `67ec8981...` |
| this PR + #4630 | 12.388s | 12.367s | 11630.847 ms | 733.97 ms | `67ec8981...` |

Readout:

- The combined stack keeps the same output hash as the matching single-PR run for both tested shapes.
- The two PRs are stack-compatible and slightly faster together.
- The stack gain is not linear: for the 93-frame 35-step request, this PR already captures most of the practical serving improvement, while #4630 adds about 4.4 ms stage improvement on top.
- #4630 is still more useful for the longer 189-frame profile shape where communication is more visible.

Recommendation: land this PR first, then rebase #4630 on top if we want the longer-sequence communication reduction. Remaining P0 work is still layout/copy cleanup and host-sync cleanup; those dominate the remaining gap to the A-framework reference.
